src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -169,42 +171,105 @@ void NnetDecodableBase::GetOutputForFrame(int32 subsampled_frame,
+    ivector->Resize(1, ivector_->Dim(),kUndefined);
+    ivector->Row(0).CopyFromVec(*ivector_);
     return;
   } else if (online_ivector_feats_ == NULL) {
     return;
   }
+  const int32 interval = opts_.ivector_interval;
   KALDI_ASSERT(online_ivector_period_ > 0);
   // frame_to_search is the frame that we want to get the most recent iVector
   // for.  We choose a point near the middle of the current window, the concept
   // being that this is the fairest comparison to nnet2.   Obviously we could do
   // better by always taking the last frame's iVector, but decoding with
   // 'online' ivectors is only really a mechanism to simulate online operation.
-  int32 frame_to_search = output_t_start + num_output_frames / 2;
-  int32 ivector_frame = frame_to_search / online_ivector_period_;
-  KALDI_ASSERT(ivector_frame >= 0);
-  if (ivector_frame >= online_ivector_feats_->NumRows()) {
-    int32 margin = ivector_frame - (online_ivector_feats_->NumRows() - 1);
-    if (margin * online_ivector_period_ > 50) {
-      // Half a second seems like too long to be explainable as edge effects.
-      KALDI_ERR << "Could not get iVector for frame " << frame_to_search
-                << ", only available till frame "
-                << online_ivector_feats_->NumRows()
-                << " * ivector-period=" << online_ivector_period_
-                << " (mismatched --ivector-period?)";
+  if (interval == 0) { // single ivector case
+    int32 frame_to_search = output_t_start + num_output_frames / 2;
+    int32 ivector_frame = frame_to_search / online_ivector_period_;
+    KALDI_ASSERT(ivector_frame >= 0);
+    if (ivector_frame >= online_ivector_feats_->NumRows()) {
+      int32 margin = ivector_frame - (online_ivector_feats_->NumRows() - 1);
+      if (margin * online_ivector_period_ > 50) {
+        // Half a second seems like too long to be explainable as edge effects.
+        KALDI_ERR << "Could not get iVector for frame " << frame_to_search
+                  << ", only available till frame "
+                  << online_ivector_feats_->NumRows()
+                  << " * ivector-period=" << online_ivector_period_
+                  << " (mismatched --ivector-period?)";
+      }
+      ivector_frame = online_ivector_feats_->NumRows() - 1;
+    }
+    ivector->Resize(1, online_ivector_feats_->NumCols(), kUndefined);
+    ivector->Row(0).CopyFromVec(online_ivector_feats_->Row(ivector_frame));
+  } else { // multiple ivectors case
+    int32 left_context = nnet_left_context_ + opts_.extra_left_context,
+          right_context = nnet_right_context_ + opts_.extra_right_context;
+    // num_ivectors_left is the num of ivectors for frames whose "t" index < 0.
+    // num_ivectors_right is the num of ivectors for frames whose "t" index >= 0
+    // They are computed according to how Round descriptor works, which
+    // basically returns floor(t / <t-modulus>) * <t-modulus>.
+    // (The assumption is that the t index of the first outptut frame is 0)
```

Comment:
Thanks for your suggestion.

Yiming

On Thursday, February 4, 2016, Daniel Povey <notifications@github.com>
wrote:

> In src/nnet3/nnet-am-decodable-simple.cc
> <https://github.com/kaldi-asr/kaldi/pull/468#discussion_r51834039>:
>
> > +                  << online_ivector_feats_->NumRows()
> > +                  << " * ivector-period=" << online_ivector_period_
> > +                  << " (mismatched --ivector-period?)";
> > +      }
> > +      ivector_frame = online_ivector_feats_->NumRows() - 1;
> > +    }
> > +    ivector->Resize(1, online_ivector_feats_->NumCols(), kUndefined);
> > +    ivector->Row(0).CopyFromVec(online_ivector_feats_->Row(ivector_frame));
> > +  } else { // multiple ivectors case
> > +    int32 left_context = nnet_left_context_ + opts_.extra_left_context,
> > +          right_context = nnet_right_context_ + opts_.extra_right_context;
> > +    // num_ivectors_left is the num of ivectors for frames whose "t" index < 0.
> > +    // num_ivectors_right is the num of ivectors for frames whose "t" index >= 0
> > +    // They are computed according to how Round descriptor works, which
> > +    // basically returns floor(t / <t-modulus>) * <t-modulus>.
> > +    // (The assumption is that the t index of the first outptut frame is 0)
>
> Regarding this function GetCurrentIvector...
> I think it would make sense to write a modified version of the function
> called GetCurrentIvector that outputs to VectorBase, and instead of taking
> int32 output_t_start, int32 num_output_frames
> to specify the time, it just takes one t value
> int32 t;
> (and it will find the closest iVector to the time specified)
> and when you want to get the matrix of iVectors you call it repeatedly.
> Of course if you need you could write a function to get the matrix of
> iVectors, that
> would call this function.
> Anyway, this is just my idea of how to refactor this to make it cleaner.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/468/files#r51834039>.
>


-- 
Sent from my iPhone